scripts/core/EXT_EventQueryKernelTimestamps.rst

```diff
@@ -0,0 +1,79 @@
+<%
+import re
+from templates import helper as th
+%><%
+    OneApi=tags['$OneApi']
+    x=tags['$x']
+    X=x.upper()
+%>
+:orphan:
+
+.. _ZE_extension_event_query_kernel_timestamps:
+
+=========================================
+ Event Query Kernel Timestamps Extension
+=========================================
+
+API
+----
+
+* Functions
+
+
+    * ${x}EventQueryKernelTimestampsExt
+
+
+* Enumerations
+
+
+    * ${x}_event_query_kernel_timestamps_ext_version_t
+
+
+Event Query Kernel Timestamps
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This extension enables the query of event timestamps to provide up to three views of the data, for a given event on a given device:
+
+1. Host timestamps, normalized with Device timestamps
+2. Device timestamps, normalized with Host timestamps
+3. Raw device timestamps
+
+Each view requires unique backing, provided by the caller.
+
+**Notes**
+
+- There is a slight performance implication to consider when requesting normalized timestamp data as the runtime must perform some calculations on behalf of the caller.
+- This extension is designed to complement and eventually replace all usages of ${x}EventQueryTimestampsExp and ${x}EventQueryKernelTimestamp.
+
+.. parsed-literal::
+
+    // Storage for normalized Host timestamps that correlate with Device timestamps, in nanoseconds
+    std::vector<uint64_t> hostTimestamps;
+
+    // Storage for normalized Device timestamps that correlate with Host timestamps, in nanoseconds
+    std::vector<uint64_t> deviceTimestamps;
+
+    // Storage for raw device timestamps
+    std::vector<${x}_kernel_timestamp_result_t> rawDeviceTimestamps;
+
+    // ...
+    // launch kernel
+    // synchronize host
+    // ...
+
+    // Number of event timestamps
+    uint32_t count = 0;
+
+    // Get the number of timestamps associated with the event.
+    ${x}EventQueryKernelTimestampsExt(hEvent, hDevice, &count, nullptr, nullptr, nullptr);
+
+    // Allocate storage for timestamp data
+    hostTimestamps.resize(count);
+    deviceTimestamps.resize(count);
+    rawDeviceTimestamps.resize(count);
+
+    // Query the event timestamps
+    ${x}EventQueryKernelTimestampsExt(hEvent, hDevice, &count,
+        hostTimestamps.data(),          /\* [optional] normalized event host timestamp data that correlates to device timestamps \*/
```

Comment:
I would like to first have a dicsussion on how we are expected to implement this in Level Zero.
Right now we cannot easily transfer device timestamps to host timestamp, there is no easy mapping.
In OCL to do this, we need to query submit host timestamps and then compute delta's.
Querying host timestamp is massive cost as it requires going to KMD.
With current append call taking ~2us, this KMD trip is 7us, so call is 4x slower, this wouldn't be acceptable for our clients profiling their workloads. (too much overhead drastically changing the outcome of workload)

With this extension we would need to go to KMD for normal timestamp events , so we could service device -> host transition later, this would introduce humongous overhead that is not present today.

Reply:
thanks @MichalMrozek . Performance hit is mentioned above "There is a slight performance implication to consider when requesting normalized timestamp data as the runtime must perform some calculations on behalf of the caller."

Main objective here is to have parity with OCL, where users dont need to do all calculations to try to map to the host, so we would have parity on both API and implementation (if L0 would have overhead, it should be the same as OCL).

Because of the performance considerations, this API (and the other ones we already have) still return raw timestamps for applications that don't want to have that overhead and prefer to access the raw timestamps.

what do you think?

Reply:
@MichalMrozek could you elaborate on why would we tax all timestamps events? As I see, as long as users dont call this API, then all timestamps obtained with already-existing APIs would have same overhead, correct?

Reply:
how would you then service this API and provide host timestamps for device timestamps?

Reply:
@MichalMrozek : could you elaborate on the question? I still think that any additional overhead that you mention, like extra calls to KMD, would be specific to this call, or isn't that the case?

Reply:
If you call to KMD on this call, you receive current host & device pair of timestamps that is after the execution happened.

You cannot correlate those to device timestamps that you have recorded when the kernel was executed as you do not know how many times device timestamp (which is 32 bit) overflowed (it does it every 343 seconds), so you need to have host time recorded close to the time when event is signaled.

Reply:
thanks @MichalMrozek . is that the same thing OpenCL GPU driver implementation is doing?

is your suggestion then to have a new flag, something like ZE_EVENT_POOL_FLAG_KERNEL_MAPPED_TIMESTAMP, so we know ahead of time whether overhead is needed?

Reply:
correct, this way applications that want to pay the full price can do this without affecting applications that doesn't want to pay for this.